### ISSUE TITLE
If a class tries to extend a Test::Class::Moose class that doesn't exist, this is silently ignored and no error is generated

### ISSUE THREAD

**Issue by skington** [houseabsolute/test-class-moose#70]
Consider the following toy example, based on a bug I just hit after an incomplete refactor in our existing code base:

``` perl
#!/usr/bin/env perl

use strict;
use warnings;
no warnings 'uninitialized';

use Test::Class::Moose::Runner;

package Parent {
    use Test::Class::Moose;

    sub test_startup {
        my ($self) = @_;
        $self->next::method;

        diag 'Parent class started up';
    }
    sub test_shutdown {
        my ($self) = @_;
        $self->next::method;

        diag 'Parent class shut down';
    }
}

package Sibling {
    ### FIXME: this is the bit that causes the class to vanish
    use Test::Class::Moose extends => 'Class::That::Does::Not::Exist';

    sub test_startup {
        my ($self) = @_;
        $self->next::method;

        diag 'Sibling class started up';
    }
    sub test_shutdown {
        my ($self) = @_;
        $self->next::method;

        diag 'Sibling class shut down';
    }
}

package Child {
    use Test::Class::Moose extends => 'Parent';
    use parent 'Sibling';
    sub test_startup {
        my ($self) = @_;
        $self->next::method;

        diag 'Child class started up';
    }

    sub test_shutdown {
        my ($self) = @_;
        $self->next::method;

        diag 'Child class shut down';
    }

    sub test_something {
        ok('We ran a test!');
    }
}

my $runner = Test::Class::Moose::Runner->new;
$runner->runtests;
```

If you run this, you get

```
1..2
ok 1 - Child {
    # Parent class started up
    # Child class started up
    1..1
    ok 1 - test_something {
        ok 1
        1..1
    }
    # Parent class shut down
    # Child class shut down
}
ok 2 - Parent {
    1..0 # SKIP Skipping 'Parent': no test methods found
}
```

If you remove the extends statement in the Sibling package, so it just says `use Test::Class::Moose`, you get what was intended:

```
1..3
ok 1 - Child {
    # Sibling class started up
    # Parent class started up
    # Child class started up
    1..1
    ok 1 - test_something {
        ok 1
        1..1
    }
    # Sibling class shut down
    # Parent class shut down
    # Child class shut down
}
ok 2 - Parent {
    1..0 # SKIP Skipping 'Parent': no test methods found
}
ok 3 - Sibling {
    1..0 # SKIP Skipping 'Sibling': no test methods found
}
```

I don't know what the best approach is here, whether Test::Class::Moose should squawk that Class::That::Does::Not::Exist is not a Test::Class::Moose class, or just ignore that bit and still import Test::Class::Moose, but it shouldn't silently do nothing.

**Comment by autarch** [comment]
I started digging into this. It looks like this behavior is pretty similar to what Moose itself does internally. Moose doesn't actually check if the class you're extending exists (to my surprise). I'm not sure what TCM should do here. I think Moose's theory is that the class might _eventually_ exist. OTOH, this seems much more likely to hurt you than help you, since I suspect most of the time you've typoed the class, and you'd be happier to get an error.

**Comment by autarch** [comment]
Hmm, I take it back. It looks like Moose will eventually cause an error to be thrown if you write `extends 'Does::Not::Exist'`, though it's through a somewhat odd chain of code. So now I'm really confused why this is happening.

**Comment by HaraldJoerg** [comment]
I found this issue as an interesting challenge for my [CPAN-PRC](http://cpan-prc.org/) assignment.  So far, I've found this:

- It doesn't matter whether the base class _exists._  To trigger the error it is sufficient if the base class does not use Test::Class::Moose itself: In that case, the extending class (Sibling in the example) will no longer be considered eligible to contain tests because TCM is not included in its list of superclasses in TCM::import.
- In Moose, if you extend a non-existing class you get a warning `Can't locate package Class::That::Does::Not::Exist for @%s::ISA`, which comes from Perl's `UNIVERSAL::isa`.   In TCM, there is no such warning because `isa` is being resolved as `Test::Deep::isa`.
- Moo dies if you extend a non-existing package.  It checks the symbol table to find out whether a package has been loaded e.g. from a module with a different name.

The documentation of TCM states that the "extends" keyword in the import list of TCM is designed for inheriting from other TCM classes.  I suggest that if the base class fails to be a TCM class, then TCM should die with a message like `'Class::That::Does::Not::Exist' is no Test::Class::Moose class at t/lib/Sibling.pm line 4.`  The author of Sibling.pm will know whether he made an error in the package name, or forgot to use TCM in the base class.

I'll prepare a pull request for that purpose in the next days.
